### PR TITLE
821 self recharging wallet srw

### LINKED ISSUE #821: Self-Recharging Wallet (SRW)
*Validators had 1 ETH in their wallet,
*Wallet is supposed to self charge from SRW but it reduced to 0.8.
*Now they have to top up manually each wallet, and is it's expensive. 

{will be adding more info}

### PR DESCRIPTION
**Pull request by yavrsky**

### CODE REVIEW

**Comment by yavrsky** [comment]
recheck

**Comment by yavrsky** [comment]
I have read the CLA Document and I hereby sign the CLA

File: contracts/Wallets.sol [hunk 1]
```diff
@@ -72,25 +74,21 @@ contract Wallets is Permissions, IWallets {
     function refundGasByValidator(
         uint validatorId,
         address payable spender,
-        uint spentGas
+        uint gasLimit
     )
         external
         override
-        allowTwo("SkaleManager", "SkaleDKG")
+        allow("SkaleManager")
```
Comment (payvint) [inline contracts/Wallets.sol:116]: Why did you remove SkaleDKG contract? It does not use this function?

File: package.json [hunk 2]
```diff
@@ -42,9 +42,9 @@
   "dependencies": {
     "@nomiclabs/hardhat-ethers": "^2.1.0",
     "@openzeppelin/contracts": "^4.7.1",
-    "@openzeppelin/contracts-upgradeable": "^4.4.2",
+    "@openzeppelin/contracts-upgradeable": "^4.6.0",
     "@openzeppelin/hardhat-upgrades": "^1.13.0",
-    "@skalenetwork/skale-manager-interfaces": "^1.0.0-develop.3",
+    "@skalenetwork/skale-manager-interfaces": "1.0.0-srw-interfaces.4",
```
Comment (payvint) [inline package.json]: Change to develop version

**Review by payvint — DISMISSED** [review]

File: contracts/ConstantsHolder.sol [hunk 3]
```diff
@@ -128,6 +128,8 @@ contract ConstantsHolder is Permissions, IConstantsHolder {
 
     uint public complaintTimeLimit;
 
+    uint public maxNodeDeposit;
```
Comment (DimaStebaev) [inline contracts/ConstantsHolder.sol]: Why `maxNodeDeposit` but not `minNodeDeposit`?
Maybe `balance` is better than `deposit` in this context?

File: contracts/SkaleManager.sol [hunk 4]
```diff
@@ -106,7 +106,7 @@ contract SkaleManager is IERC777Recipient, ISkaleManager, Permissions {
     }
 
     function nodeExit(uint nodeIndex) external override {
-        uint gasTotal = gasleft();
+        uint gasLimit = (gasleft() + 7600) * 64 / 63 + 21000;
```
Comment (DimaStebaev) [inline contracts/SkaleManager.sol]: This calculations is duplicated. To my thinking it worth moving them to private function.

Also I don't like that raw numbers are used. I think it's better to create constants. It improves readability.

`gasLimit = (gasleft() + HEADER_COSTS) * CALL_GAS_MULTIPLIER + CALL_PRICE` is easier to understand.

File: contracts/SkaleManager.sol [hunk 5]
```diff
@@ -149,7 +151,7 @@ contract SkaleManager is IERC777Recipient, ISkaleManager, Permissions {
     }
 
     function getBounty(uint nodeIndex) external override {
-        uint gasTotal = gasleft();
+        uint gasLimit = (gasleft() + 7600) * 64 / 63 + 21000;
```
Comment (DimaStebaev) [inline contracts/SkaleManager.sol]: See my comment above.

File: contracts/SkaleManager.sol [hunk 6]
```diff
@@ -196,10 +200,4 @@ contract SkaleManager is IERC777Recipient, ISkaleManager, Permissions {
             "Token was not minted"
         );
     }
-
-    function _refundGasByValidator(uint validatorId, address payable spender, uint spentGas) private {
```
Comment (DimaStebaev) [inline contracts/SkaleManager.sol:200]: Why was this function removed? The removal causes code duplication.

File: dictionaries/domain-terms.txt [hunk 7]
```diff
@@ -13,4 +13,5 @@ reentrancy
 secp256k1
 untrusted
 Timelimit
+impls
```
Comment (DimaStebaev) [inline dictionaries/domain-terms.txt]: Doubtful.

File: migrations/upgrade.ts [hunk 8]
```diff
@@ -25,22 +27,52 @@ export async function setNewVersion(safeTransactions: string[], abi: SkaleABIFil
     ));
 }
 
+function getContractsWithout(contractName: string | undefined): string[] {
+    if (!contractName)
```
Comment (DimaStebaev) [inline migrations/upgrade.ts]: Braces are too expensive to waste them.

File: migrations/upgrade.ts [hunk 9]
```diff
@@ -25,22 +27,52 @@ export async function setNewVersion(safeTransactions: string[], abi: SkaleABIFil
...
 async function main() {
     await upgrade(
         "skale-manager",
         "1.9.2",
         getDeployedVersion,
         setNewVersion,
-        // Permissions in SkaleManager are needed to set a new version
         ["SkaleManager"],
-        ["ContractManager"].concat(contracts),
+        getContractsWithout("ConstantsHolder"), // Remove ConstantsHolder from contracts to do upgradeAndCall
```
Comment (DimaStebaev) [inline migrations/upgrade.ts:48]: Let it be but I prefer to modify `upgrade-tools` to support such things.

File: scripts/change_manifest.py [hunk 10]
```diff
@@ -0,0 +1,18 @@
+import sys
+import json
+
+
+def main():
```
Comment (DimaStebaev) [inline scripts/change_manifest.py:9]: This script corrupts a manifest if it's original size is bigger than processed one.

Fix the bug or replace with already fixed version from here:
https://github.com/skalenetwork/IMA/blob/08ed0cc81ea48661e0af4614de21891da05fee15/proxy/scripts/updateManifest.py

File: scripts/test_upgrade.sh [hunk 11]
```diff
@@ -41,6 +41,7 @@ cd $GITHUB_WORKSPACE
 nvm use $CURRENT_NODE_VERSION
 rm -r --interactive=never $DEPLOYED_DIR
 
+python3 scripts/change_manifest.py $GITHUB_WORKSPACE/.openzeppelin/unknown-*.json
```
Comment (DimaStebaev) [inline scripts/test_upgrade.sh:45]: Add comment to remove this line after next upgrade.

File: test/SkaleManager.ts [hunk 12]
```diff
@@ -215,7 +224,7 @@ describe("SkaleManager", () => {
                     getPublicKey(nodeAddress), // public key
                     "d2", // name
                     "some.domain.name");
-                await wallets.rechargeValidatorWallet(validatorId, {value: 1e18.toString()});
+                await wallets.rechargeValidatorWallet(validatorId, {value: 5e18.toString()});
```
Comment (DimaStebaev) [inline test/SkaleManager.ts]: `ethers.utils.parseEther`

File: contracts/SkaleManager.sol (hunk 4 above)
Comment (payvint) [reply contracts/SkaleManager.sol]: Hmm maybe it would be better to have a private function `_gasLimitCalculation` without moving teach individual value to constant

File: migrations/upgrade.ts (hunk 8 above)
Comment (payvint) [reply migrations/upgrade.ts]: :money_with_wings:

**Review by DimaStebaev — APPROVED** [review]

**Review by gannakulikova — APPROVED** [review]
